src/type.js

```diff
@@ -0,0 +1,199 @@
+}
+
+/**
+ * Validates the instance of a value against a specific class.
+ * @param {!Object} expectedClass Class to check instance against.
+ * @return {function} Type validator.
+ */
+function createInstanceTypeValidator(expectedClass) {
+	return (stateValue, stateKey) => {
+		if (!(stateValue instanceof expectedClass)) {
+			return `Expected instance of ${expectedClass}`;
+		}
+
+		return null;
+	}
+}
+
+/**
+ * Validates each value in an object against a certain type.
+ * @param {!function} expectedType Type validator.
+ * @return {function} Object validator.
+ */
+function createObjectOfValidator(expectedType) {
+	return (stateValue, stateKey) => {
+		let success = true;
+
+		for (let key in stateValue) {
+			success = expectedType(stateValue[key], null, true) === null;
+		}
+
+		if (!success) {
+			return 'Expected object of one type';
+		}
+
+		return null;
+	}
+}
+
+/**
+ * Validates an objects shape.
+ * @param {!Object} shapeTypes An object containing type validators for each key.
+ * @return {function} Shape validator.
+ */
+function createShapeTypeValidator(shapeTypes) {
+	return (stateValue, stateKey) => {
+		let success = true;
+
+		for (let key in shapeTypes) {
+			const validator = shapeTypes[key];
+
+			if (!validator) {
+				continue;
+			}
+
+			const stateValueForKey = stateValue[key];
+
+			if (stateValueForKey) {
+				const success = validator(stateValueForKey, null, true) === null;
+
+				if (!success) {
+					return 'Expected object with a specific shape';
+				}
+			}
+		}
+
+		return null;
+	}
+}
+
+/**
+ * Validates a specific primitive type.
+ * @param {!string} expectedType Name of primitive type.
+ * @return {function} Type validator.
+ */
+function createPrimitiveTypeValidator(expectedType) {
+	return (stateValue, stateName) => {
+		const stateType = getStateType(stateValue);
+
+		if (stateType !== expectedType) {
+			return `Expected type ${expectedType}`;
+		}
+		return null;
+	}
+}
+
+/**
+ * Validates if any of the type validators are successfull.
+ * @param {!array} arrayOfTypeValidators Array of type validators.
+ * @return {string|function} Type validator or error.
+ */
+function createUnionTypeValidator(arrayOfTypeValidators) {
+	if (!Array.isArray(arrayOfTypeValidators)) {
+		return 'Expected an array';
+	}
+
+	return (stateValue, stateName) => {
+		for (var i = 0; i < arrayOfTypeValidators.length; i++) {
+			var validator = arrayOfTypeValidators[i];
+
+			if (validator(stateValue, stateName, true) == null) {
+				return null;
+			}
+		}
+
+		return 'Expected one of given types.';
+	}
+}
+
+/**
+ * Wrapper function that is used to determine if console.error should be thrown.
+ * @param {!function} validator Type validator.
+ * @return {function} This is the result function passed to the state validator.
+ */
+function emitWarning(validator) {
+	function returnFn(stateValue, stateName, disableWarning) {
+		const expectedError = validator(stateValue, stateName);
+
+		if (!disableWarning && expectedError) {
```

Comment:
The `disableWarning` argument is primarily for nested validators, such as `shape` or `oneOfType`.

```js
		typeValidators.oneOfType(
			[
				typeValidators.string,
				typeValidators.number
			]
		);
```

We disable the warning so that when we iterate through each validator, we would only get one console.error.

In the validator above, if we pass in a string, we disable the warning when it gets to the `typeValidators.number` validator.

Does that make sense? I may need to refactor it a bit to make more logical sense for the `emitWarning` function.